lastlog.8
.\" Copyright 1992, Phillip Street and John F. Haugh II
.\" All rights reserved.
.\"
.\" Permission is granted to copy and create derivative works for any
.\" non-commercial purpose, provided this copyright notice is preserved
.\" in all copies of source code, or included in human readable form
.\" and conspicuously displayed on all copies of object code or
.\" distribution media.
.\"
.\"	@(#)lastlog.8	3.2	10:35:49	18 Jul 1993 (National Guard Release)
.\"
.TH LASTLOG 8
.SH NAME
lastlog \- examine lastlog file
.SH SYNOPSIS
/etc/lastlog [ -u uid ] [ -t days ]
.SH DESCRIPTION
\fIlastlog\fR formats and prints the contents of the last login log,
\fI/usr/adm/lastlog\fR.  The \fBlogin-name\fR, \fBport\fR, and \fBlast login 
time\fR will be printed.
The default (no flags) causes lastlog entries to be printed in UID
order.
Entering \fB-u login-name\fR flag will
cause the lastlog record for \fBlogin-name\fR only to be printed.
Entering \fB-t days\fR will cause only the
lastlogins more recent than \fBdays\fR to be printed.
The \fB-t\fR flag overrides the use of \fB-u\fR.
.PP
If the user has never logged in the message \fB"**Never logged in**"\fR will 
be displayed instead of the port and time.
.SH Files
/usr/adm/lastlog \- lastlog logging file
.SH Caveats
Large gaps in uid numbers will cause the lastlog program to run longer with
no output to the screen (i.e. if mmdf=800 and last uid=170, program will
appear to hang as it processes uid 171-799).


NEWS
$Id$

shadow-4.1.5.1 -> shadow-4.2					UNRELEASED

*** general
  * Handle libc whose crypt() returns NULL when passed a salt that
    violates specs or system requirements (e.g. FIPS140). This is needed
    with glibc/eglibc 2.17 for tools checking passwords (passwd (non PAM
    enabled) or newgrp), and for tools generating encrypted passwords
    (chgpasswd, chpasswd, or gpasswd when non PAM enabled or when a fixed
    crypt method is requested on the command line, and newusers, or passwd
    in their non PAM enabled versions)
  * Fix segfault when reading groups split on multiple lines. This impacts
    most user/group management tools when MAX_MEMBERS_PER_GROUP is set.

- su
  * When su receives a signal (SIGTERM, or SIGINT/SIGQUIT in non
    interactive mode), kill the child process group, rather than just the
    immediate child.
  * Fix segmentation faults for users without a proper home or shell in
    their passwd entries.

- login
  * Fix segmentation faults for users without a proper home or shell in
    their passwd entries.

*** documentation
  * Fixed useradd man page (--home-dir option, instead of --home).

*** translation
  * Updated Russian translation.
  * Updated German man pages translation.
  * Fixed gshadow Japanese man page translation.

shadow-4.1.5 -> shadow-4.1.5.1					2012-05-25

- login
  * Log into utmp(x) when PAM is enabled, but do not log into wtmp.
    This complete pam_lastlog which logs into wtmp and in into utmp(x).
- su
  * non PAM enabled versions: do not fail if su is called without a
    controlling terminal.
- userdel
  * Fix segfault when userdel removes the user's group.

*** documentation
  * .so links now point to paths relative to the top-level manual hierarchy

*** translation
  * Updated French man pages translation.
  * Updated German man pages translation.
  * Updated Polish man pages translation. (logoutd.8)

shadow-4.1.4.3 -> shadow-4.1.5					2012-02-12

*** security
  * su -c could be abused by the executed command to invoke commands with
    the caller privileges. See below. (CVE-2005-4890)

*** general
  * report usage error to stderr, but report usage help to stdout (and return
    zero) when explicitly requested (e.g. with --help).
  * initial support for tcb (http://openwall.com/tcb/) for useradd,
    userdel, usermod, chage, pwck, vipw.
  * Added support for ACLs and Extended Attributes in useradd and usermod.
    Support shall be enabled with the new --with-acl or --with-attr
    configure options.
  * Added diagnosis for lock failures.
  * use libsemanage instead of the semanage tool.

- chage
  * Add --root option.
- chfn
  * Add --root option.
- chgpasswd
  * When the gshadow file exists but there are no gshadow entries, an entry
    is created if the password is changed and group requires a
    shadow entry.
  * Add --root option.
- chpasswd
  * PAM enabled versions: restore the -e option to allow restoring
    passwords without knowing those passwords. Restore together the -m
    and -c options. (These options were removed in shadow-4.1.4 on PAM
    enabled versions)
  * When the shadow file exists but there are no shadow entries, an entry
    is created if the password is changed and passwd requires a
    shadow entry.
  * Add --root option.
- chsh
  * Add --root option.
- faillog
  * The -l, -m, -r, -t options only act on the existing users, unless -a is
    specified.
  * Add --root option.
- gpasswd
  * Add --root option.
- groupadd
  * Add --root option.
- groupdel
  * Add --root option.
- groupmems
  * Fix parsing of gshadow entries.
  * Add --root option.
- groupmod
  * Fixed groupmod when configured with --enable-account-tools-setuid.
  * When the gshadow file exists but there are no gshadow entries, an entry
    is created if the password is changed and group requires a
    shadow entry.
  * Add --root option.
- grpck
  * Add --root option.
  * NIS entries were dropped by -s (sort).
- grpconv
  * Add --root option.
- grpunconv
  * Add --root option.
- lastlog
  * Add --root option.
- login
  * Fixed limits support (non PAM enabled versions only)
  * Added support for infinite limits and group based limits (non PAM
    enabled versions only)
  * Fixed infinite loop when CONSOLE is configured with a colon-separated
    list of TTYs.
  * Fixed warning and support for CONSOLE_GROUPS for users member of more
    than 16 groups.
  * Do not log into utmp(x) or wtmp when PAM is enabled. This is done by
    pam_lastlog.
- newgrp, sg
  * Fix parsing of gshadow entries.
- newusers
  * Add --root option.
- passwd
  * Add --root option.
- pwpck
  * NIS entries were dropped by -s (sort).
  * Add --root option.
- pwconv
  * Add --root option.
- pwunconv
  * Add --root option.
- useradd
  * If the skeleton directory contained hardlinked files, copies of the
    hardlink were removed from the skeleton directory.
  * Add --root option.
- userdel
  * Check the existence of the user's mail spool before trying to remove
    it. If it does not exist, a warning is issued, but no failure.
  * Do not remove a group with the same name as the user (usergroup) if
    this group isn't the user's primary group.
  * Add --root option.
  * Add --selinux-user option.
- usermod
  * Accept options in any order (username not necessarily at the end)
  * When the shadow file exists but there are no shadow entries, an entry
    is created if the password is changed and passwd requires a
    shadow entry, or if aging features are used (-e or -f).
  * Add --root option.
- su
  * Document the su exit values.
  * When su receives a signal, wait for the child to terminate (after
    sending a SIGTERM), and kill it only if it did not terminate by itself.
    No delay will be enforced if the child cooperates.
  * Default ENV_SUPATH is /sbin:/bin:/usr/sbin:/usr/bin
  * Fixed infinite loop when CONSOLE is configured with a colon-separated
    list of TTYs.
  * Fixed warning and support for CONSOLE_GROUPS for users member of more
    than 16 groups.

lastlog.c
/*
 * Copyright 1989, 1990, 1992, John F. Haugh II
 * All rights reserved.
 *
 * Permission is granted to copy and create derivative works for any
 * non-commercial purpose, provided this copyright notice is preserved
 * in all copies of source code, or included in human readable form
 * and conspicuously displayed on all copies of object code or
 * distribution media.
 *
 * Modified from faillog.c by Philip Street
 * Installation:  Maine National Guard
 */

#include <sys/types.h>
#include <sys/stat.h>
#include <stdio.h>
#include "pwd.h"
#include <time.h>
#ifndef	BSD
#include <string.h>
#include <memory.h>
#else
#include <strings.h>
#define	strchr	index
#define	strrchr	rindex
#endif

#define LASTFILE "/usr/adm/lastlog"

#include "config.h"
#include "lastlog.h"

#ifndef	lint
static	char	sccsid[] = "@(#)lastlog.c	3.2	10:30:33	18 Jul 1993";
#endif

FILE	*lastlogfile;	/* lastlog file stream */
off_t	user;		/* one single user, specified on command line */
int	days;		/* number of days to consider for print command */
time_t	seconds;	/* that number of days in seconds */

int	uflg;		/* set if user is a valid user id */
int	tflg;		/* print is restricted to most recent days */
struct	lastlog	lastlog; /* scratch structure to play with ... */
struct	stat	statbuf; /* fstat buffer for file size */

int freadcode;

extern	int	optind;
extern	char	*optarg;
extern	char	*asctime ();
extern	struct	passwd	*getpwuid ();
extern	struct	passwd	*getpwnam ();
extern	struct	passwd	*getpwent ();
extern	struct	tm	*localtime ();

#define	DAY	(24L*3600L)
#define	NOW	(time ((time_t *) 0))

main (argc, argv)
int	argc;
char	**argv;
{
	int	c;
	struct	passwd	*pwent;

	if ((lastlogfile = fopen (LASTFILE,"r")) == (FILE *) 0) {
		perror (LASTFILE);
		exit (1);
	}
	while ((c = getopt (argc, argv, "u:t:")) != EOF) {
		switch (c) {
			case 'u':
				pwent = getpwnam (optarg);
				if (! pwent) {
					fprintf (stderr, "Unknown User: %s\n", optarg);
					exit (1);
				}
				uflg++;
				user = pwent->pw_uid;
				break;
			case 't':
				days = atoi (optarg);
				seconds = days * DAY;
				tflg++;
				break;
		}
	}
	print ();
	fclose (lastlogfile);
	exit (0);
	/*NOTREACHED*/
}

print ()
{
	int	uid;
	off_t	offset;

	if (uflg) {
		offset = user * sizeof lastlog;
		fstat (fileno (lastlogfile), &statbuf);
		if (offset >= statbuf.st_size)
			return;

		fseek (lastlogfile, (off_t) user * sizeof lastlog, 0);
		if (fread ((char *) &lastlog, sizeof lastlog, 1, lastlogfile) == 1)
			print_one (user);
		else
			perror (LASTFILE);
	} else {
		for (uid = 0;
			fread ((char *) &lastlog, sizeof lastlog, 1, lastlogfile) == 1;
				uid++) {

			if (tflg && NOW - lastlog.ll_time > seconds)
				continue;

			print_one (uid);
		}
	}
}

print_one (uid)
int	uid;
{
	static	int	once;
	char	*cp;
	struct	tm	*tm;
	struct	passwd	*pwent;

	if (! once) {
		printf ("Username\t\tPort\tLatest\n");
		once++;
	}
	pwent = getpwuid (uid);
	tm = localtime (&lastlog.ll_time);
	cp = asctime (tm);
	cp[24] = '\0';

	if(lastlog.ll_time == (time_t) 0)
		cp = "**Never logged in**\0";

	if (pwent) {
		printf ("%-16s\t%s\t%s\n",pwent->pw_name,lastlog.ll_line,cp);
	}
}
